### ISSUE TITLE
add `no-arrow-function-as-method` rule

### ISSUE THREAD

**Issue by cafadev** [vuejs/eslint-plugin-vue#2016]
**Please describe what the rule should do:**

When we're using Option API it's clearly to see the code separation among the component sections. There is a rule called `vue/order-in-components`, in order to enhance the behavior of this rule, it would be a good idea to avoid arrow function in the root of the `setup()`/`<script setup>`. So in that way we can keep all the functions as methods, just like in Option API.

**What category should the rule belong to?**

[ X ] Enforces code style (layout)
[   ] Warns about a potential error (problem)
[   ] Suggests an alternate way of doing something (suggestion)
[   ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

```javascript
// Good
const count = ref(0)

function increment() {
  count.value += 1;
}

function resetCount() {
  count.value = 0
}

// Bad
const count = ref(0)

const increment = () => {
  count.value += 1;
}

const resetCount = () => {
  count.value = 0
}
```

**Comment by ota-meshi** [comment]
Thank you for posting issue.

I think we can do that using the `func-style` rule's `"declaration"` option.

https://eslint.org/docs/latest/rules/func-style
https://eslint.org/docs/latest/rules/func-style#declaration

[Online demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IGZ1bmMtc3R5bGU6IFtcImVycm9yXCIsIFwiZGVjbGFyYXRpb25cIl0gKi9cbi8vIEJhZFxuY29uc3QgY291bnQgPSByZWYoMClcblxuY29uc3QgaW5jcmVtZW50ID0gKCkgPT4ge1xuICBjb3VudC52YWx1ZSArPSAxO1xufVxuXG5jb25zdCByZXNldENvdW50ID0gKCkgPT4ge1xuICBjb3VudC52YWx1ZSA9IDBcbn1cblxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJmdW5jLXN0eWxlIjpbImVycm9yIl19LCJlbnYiOnt9fX0=)

**Comment by cafadev** [comment]
Yeah, but that rule prevent the use of arrow functions in all the code. Sometimes we would want to use arrow functions inside declarative functions (just as an example). The mainly different for the proposal rule is to avoid arrow functions just in the setup root level. For example:

```vue
<script lang="ts" setup>
// Good
const count = ref(0)
function increment(debounceTime=false) {

  const randomFunction = () => {
    // Do something
  }

  count.value += 1;
  randomFunction()
}

function resetCount() {
  count.value = 0
}
</script>
```

```vue
<script lang="ts">
export default defineComponent({
  setup() {
    // Good
    const count = ref(0);
    function increment(debounceTime = false) {
      const randomFunction = () => {
        // Do something
      };

      count.value += 1;
      randomFunction();
    }

    return {
      count,
      increment
    }
  }
});
</script>
```

Thanks for take the time to read this proposal 👍🏼.

**Comment by ota-meshi** [comment]
If you want to use arrow functions, I think it's good for consistency to use arrow functions everywhere.
Why do you want different function styles depending on where you use them?

> it would be a good idea to avoid arrow function in the root of the `setup()`/`<script setup>`. 

This explanation is not clear to me. What's wrong with using arrow functions specifically?

**Comment by thimonwentink** [comment]
When our team switched from Options API to Composition API, we also wanted to use function statements. It would be nice if this plugin made a rule for this.

In the meantime, I wrote a rule in our own eslint plugin at Programic. You can use it if you want: `@programic/vue-no-function-expressions`. I also wrote one for Pinia: `@programic/pinia-no-function-expressions`. Check out [the repo](https://github.com/programic/eslint-plugin).
